File: src/main/java/io/gaegul/ch01/Duck.java
package io.gaegul.ch01;

/**
 * 오리 클래스
 */
public abstract class Duck {

	public void quack() {
		System.out.println("꽥꽥~");
	}

	public boolean swim() {
		return true;
	}

	public abstract void display();
}

File: src/main/kotlin/io/gaegul/ch01/Duck.kt
package io.gaegul.ch01

import io.gaegul.ch01.behavior.FlyBehavior
import io.gaegul.ch01.behavior.QuackBehavior

/**
 * 오리 슈퍼 클래스
 */
abstract class Duck(
    private var quackBehavior: QuackBehavior,
    private var flyBehavior: FlyBehavior
) {

    fun performQuack() {
        quackBehavior.quack()
    }

    fun swim(): Boolean {
        return true
    }

    fun performFly() {
        flyBehavior.fly()
    }

    abstract fun display()
}